[frontend] Show artifact preview in UI (#2172)

eterna2 opened the pull request:
Part of #2172. This PR removes the MinioArtifactLink component and replaces it with a MinioArtifactPreview. MinioArtifactPreview takes in an Argo S3Artifact object and attempts to load and render a preview to the artifact content. If the attempts fails, no preview will be shown, but the error will be logged to console.

A screenshot of the UI is shown below:
![artifactpreview](https://user-images.githubusercontent.com/1248825/81305070-7af8f000-90b0-11ea-87ef-d765ba43f14e.png)

![artifactpreview2](https://user-images.githubusercontent.com/1248825/81468916-e52aa580-9214-11ea-86df-95d9206f4624.png)

cc: @Bobgy

Bobgy commented:
@eterna2 This looks great!

Can you add a screenshot when there are more than one artifacts?

eterna2 commented on frontend/src/lib/Utils.tsx:
```diff
@@ -292,8 +292,23 @@ export function generateGcsConsoleUri(gcsUri: string): string | undefined {
+  peek?: number,
+): string {
+  return encodeURI(
+    `artifacts/get?source=${source}&bucket=${bucket}&key=${key}${peek ? `&peek=${peek}` : ''}`,
```
i intentionally did not use buildQuery because of the encodeUriComponent. Because I think the keys and bucket etc from the S3 artifact is already encoded. 

In my prev PR, i created the consistent encode and decode because of this issue - double encoded in some cases. 

But let me make the change and try again.

Bobgy replied:
I see, if there's something blocking that, feel free to leave it for future improvements.

eterna2 replied:
ok. I did a test run, so far all the workflow artifacts and ml-metadata artifacts look fine too.

Bobgy commented:
@eterna2 note you can ignore travis tests when developing, it's kind of flaky recently. And even if you retry them, they always fail.
Pushing extra empty commits can retry them.

eterna2 commented:
ok i fixed everything.

Bobgy commented on frontend/src/components/DetailsTable.tsx:
```diff
@@ -97,18 +103,16 @@ const DetailsTable = (props: DetailsTableProps) => {
             }
           }
-          // If the value isn't a JSON object, just display it as is
+          // If value is a S3Artifact render a preview, otherwise just display it as is
```
nit: update the comment?

Bobgy commented on frontend/src/components/MinioArtifactPreview.test.tsx:
```diff
@@ -0,0 +1,139 @@
+jest.mock('../lib/Apis');
+
+describe('MinioArtifactPreview', () => {
+  const readFile = Apis.readFile as jest.Mock;
```
a type safe way to get the spy is
```
const readFileSpy = jest.spyOn(Apis, 'readFile');
```
You can remove the `jest.mock` usage.

Bobgy commented on frontend/src/components/MinioArtifactPreview.test.tsx:
```diff
@@ -0,0 +1,139 @@
+      secretKeySecret: { key: 'secretkey', optional: false, name: 'minio' },
+    };
+
+    const container = document.body.appendChild(document.createElement('div'));
+    await act(async () => {
+      render(<MinioArtifactPreview artifact={s3Artifact} />, { container });
+    });
```
This is kind of strange. Why did you need to create the div manually and wrap `render` in an `act`?

Bobgy replied:
Did you want to try something like https://github.com/kubeflow/pipelines/blob/f882f361cc705580261128e33523fe80d781b30f/frontend/src/components/PodYaml.test.tsx#L37? to wait for promise to resolve before finally rendering the container?

Bobgy commented on frontend/src/lib/Apis.ts:
```diff
@@ -205,9 +206,28 @@ export class Apis {
+   * Builds an url for the readFile API to retrieve a workflow artifact.
+   * @param props object describing the artifact (e.g. source, bucket, and key)
+   */
+  public static buildReadFileUrl(path: StoragePath, namespace?: string, peek?: number) {
```
Awesome! This looks great

Bobgy commented on frontend/src/components/DetailsTable.tsx:
```diff
@@ -48,22 +47,29 @@ export const css = stylesheet({
   title?: string;
-  valueComponent?: React.FC<{ value: S3Artifact }>;
+  valueComponent?: React.FC<ValueComponentProps<T>>;
+  [key: string]: any;
```
Thanks! this looks like a good approach.
nit: probably slightly change it to
```
valueComponentProps: { [key: string]: any };
```
to make it more focused.

Otherwise, DetailsTable will accept any key as a prop, which is not good for type checking.

eterna2 commented:
ok fixed everything.

- used `valueComponentProps` to pass the extra props to the valueComponent
- fixed the unit test with `act`
- fixed the comments
- fixed the handling of invalid s3artifact (now it return null also)
- add unit test if value is string (or other primitive) -> render as string (i.e. MinioArtifactPreview will render valid s3artifact values properly, invalid objects as null, undefined/null values as null, otherwise render as a string)

eterna2 commented:
hmmm... got a unit test error
```
FAIL src/pages/RunDetails.test.tsx (41.301s)
  ● Console
    console.warn node_modules/react-dom/cjs/react-dom.development.js:12357
      Warning: componentWillMount has been renamed, and is not recommended for use. See https://fb.me/react-unsafe-component-lifecycles for details.
```

eterna2 commented:
the strangest part is that only `RunDetails switches to inputs/outputs tab in side pane 1` shld be showing the `DetailsTable`. 

not `RunDetails opens side panel when graph node is clicked 1`

eterna2 commented:
ok i found the issue. ur recent update to the visualization tab changed the snapshot. So the unit test is testing against that. 

I have rebase to the latest master and updated the snapshot. Should work now.

Ark-kun commented on frontend/src/components/MinioArtifactPreview.tsx:
```diff
@@ -0,0 +1,102 @@
+  React.useEffect(() => {
+    let cancelled = false;
+    if (storagePath) {
+      // TODO how to handle binary data
```
Later, we can abort preview if unprintable bytes are found.

Ark-kun commented on frontend/src/components/MinioArtifactPreview.tsx:
```diff
@@ -0,0 +1,102 @@
+    if (storagePath) {
+      // TODO how to handle binary data
+      Apis.readFile(storagePath, namespace, peek + 1).then(
+        data => !cancelled && setContent(data.length > peek ? `${data.slice(0, peek)} ...` : data),
```
Later we can have more advanced logic here. For example <=10 full lines or <= 10KB whichever comes first.

P.S. The ellipsis is a nice touch (maybe later we can put it on its own line)

Bobgy replied:
+1 to putting `...` on its own line, why not do it now?

Ark-kun commented:
Thank you a lot for this PR. This really improves the experience!

Ark-kun commented on frontend/src/lib/Apis.ts:
```diff
@@ -205,9 +205,28 @@ export class Apis {
+   * @param param.key key (i.e. path) of the artifact in the bucket
+   */
+  public static buildArtifactUrl({ source, bucket, key }: StoragePath) {
+    // TODO see https://github.com/kubeflow/pipelines/pull/3725
```
For now, I've decided to put the Argo artifact info to the MLMD artifact custom property. So there will be no new URIs for now.

Ark-kun commented:
/lgtm

Bobgy commented on frontend/src/components/DetailsTable.test.tsx:
```diff
@@ -55,60 +101,502 @@ describe('DetailsTable', () => {
+            >
+              key
+            </span>
+            <div
```
nit: snapshot for ace editor is long and it didn't even include the actual content.
Recommend mocking it out like https://github.com/Bobgy/pipelines/blob/68c55e035640be29f89b8f4896d045acf6595c91/frontend/src/components/PodYaml.test.tsx#L7

eterna2 replied:
ok done

Bobgy commented on frontend/src/components/MinioArtifactPreview.test.tsx:
```diff
@@ -0,0 +1,224 @@
+        <div
+          class="root"
+        >
+          <a
+            class="link"
+            rel="noopener"
+            target="_blank"
+          />
```
All the <a/>s should have href and content, is this a bug?

eterna2 replied:
yes. I missed this too. Let me check

eterna2 replied:
found the bug. cuz I mocked the entire `APIs` module so the functions did not return anything. Removed the unneccessary mocks. Works now.

eterna2 commented:
ok made the fixes.

Also changed the props from `peek` to `maxbytes` and `maxlines`. 
`maxlines` is optional.

Bobgy commented on frontend/src/components/MinioArtifactPreview.test.tsx:
```diff
@@ -1,3 +1,8 @@
+import MinioArtifactPreview from './MinioArtifactPreview';
```
nit: Can you move imports under license declaration?

eterna2 replied:
ops. ide auto import hook not smart enuff to recognize license

eterna2 replied:
done

Bobgy commented:
/approve

Bobgy commented:
Awesome! Really useful feature!

/lgtm

Bobgy commented:
/hold
unfortunately, travis tests failed
Can you submit an additional commit with any change to retry it?

Bobgy commented:
/cc @Ark-kun 
I've seen travis tests being flaky and unretriable quite often recently. Shall we consider changing it to non required and approver/lgtmer should check it passes?

Bobgy commented:
/lgtm
/unhold

Bobgy commented:
Hmmm, it failed again. I'll help you merge this PR directly.

eterna2 commented:
I looked at the logs. The Travis build was terminated because it timeout after 10 minutes.

I think u can go increase the timeout

eterna2 commented:
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
The build has been terminated
```

Bobgy commented:
> I looked at the logs. The Travis build was terminated because it timeout after 10 minutes.
> 
> I think u can go increase the timeout

Thanks, that's interesting.
Can you give a screenshot where you saw that?

Bobgy commented:
I was looking at the error in https://travis-ci.com/github/kubeflow/pipelines/jobs/331832190#L3086.
It shows up randomly for all three tests on different python versions.

eterna2 commented:
> > I looked at the logs. The Travis build was terminated because it timeout after 10 minutes.
> > I think u can go increase the timeout
> 
> Thanks, that's interesting.
> Can you give a screenshot where you saw that?

here
https://travis-ci.com/github/kubeflow/pipelines/jobs/331472823#L3205

Bobgy commented:
@eterna2 Thanks!
If you check PR travis test history here: https://travis-ci.com/github/kubeflow/pipelines/pull_requests. All other failures seem to be the same as the one I posted above. There was only one time it showed time out. We'd focus on solving the problem in https://github.com/kubeflow/pipelines/pull/3707#issuecomment-626652201 with higher priority.
/cc @Ark-kun

Ark-kun commented:
> I've seen travis tests being flaky and unretriable quite often recently.

The Travis errors are more retriable now.
There were several issues recently:
* Travis' "No output has been received in the last 10m0s". Happens rarely, but is retriable
* Coveralls error 422. This was an unretriable error. I've removed coveralls call.
* TFX installation failures. Unretriable. Fixed on the TFX side.
* TFX installation flakyness. Somewhat retriable. See my issue https://github.com/pypa/pip/issues/8218

So, only retriable flakiness remains. Sorry for the inconvenience.

Bobgy commented:
> > I've seen travis tests being flaky and unretriable quite often recently.
> 
> The Travis errors are more retriable now.
> There were several issues recently:
> 
> * Travis' "No output has been received in the last 10m0s". Happens rarely, but is retriable
> * Coveralls error 422. This was an unretriable error. I've removed coveralls call.
> * TFX installation failures. Unretriable. Fixed on the TFX side.
> * TFX installation flakyness. Somewhat retriable. See my issue [pypa/pip#8218](https://github.com/pypa/pip/issues/8218)
> 
> So, only retriable flakiness remains. Sorry for the inconvenience.

@Ark-kun thanks for fixing them, that's great news for me.

Bobgy commented on frontend/src/components/MinioArtifactPreview.tsx:
```diff
@@ -0,0 +1,128 @@
+  // TODO need to come to an agreement how to encode artifact info inside a url
+  // namespace is currently not supported
+  const linkText = Apis.buildArtifactUrl(storagePath);
+  const artifactUrl = Apis.buildReadFileUrl(storagePath, namespace, maxbytes);
```
@eterna2 I overlooked a bug. Here, maxbytes shouldn't be set, because when we click the artifact link, we should expect user wants to see full content, is that right?

eterna2 replied:
my bad. I missed this too.
